Idempotent param

stealthycoin opened the pull request:
cc @kyleknap @JordonPhillips @jamesls

kyleknap commented on tests/functional/test_ec2.py:
```diff
@@ -0,0 +1,66 @@
+
+        # Check the ClientToken parameter exsits
+        # after injection and before the call
+        self.client.meta.events.register_first(
```
This is sort of confusing on what is happening here. For this functional test, here is what I recommend:


1) Split these tests out to two test methods: one where the user provides the parameter and one where you inject it for the user. In general if you are making two somewhat unrelated assertions, it is a sign you should break them up into separate tests because if the first assertion fails you will not be able to see if the second assertion fails as well.
2) For the case where you are injecting the idempotency token for the user, I would recommend instead of registering that lambda event, use ``botocore.stub.ANY``
3) For stubbers, feel free to use the context manager to activate and deactivate the stubber.

stealthycoin replied:
1) Sounds good
2) I tried that first and it threw fit because I think the token gets inserted after the stubber looks at it? I'm not sure I can try that way again and fiddle with it more. I just figured the stubber didn't support what I wanted to do so I went around it.
3) Will do.

JordonPhillips replied:
Personally I prefer the context manager. It was added later on, which is why it isn't used in some tests.

kyleknap commented on botocore/handlers.py:
```diff
@@ -109,6 +110,24 @@ def decode_console_output(parsed, **kwargs):
             logger.debug('Error decoding base64', exc_info=True)
 
 
+def generate_idempotent_uuid(params, model, **kwargs):
```
It would be great to have a unit test here. Mainly to test that this function properly injects a uuid. You would not have to guess the exact value of the uuid but make sure it looks like an uuid.

stealthycoin commented on botocore/handlers.py:
```diff
@@ -764,6 +784,8 @@ def _replace_content(self, section):
     ('after-call.cloudformation.GetTemplate', json_decode_template_body),
     ('after-call.s3.GetBucketLocation', parse_get_bucket_location),
 
+    ('before-parameter-build.*.*', generate_idempotent_uuid),
```
Hmm without the *.* it didn't get run as far as I could tell.

stealthycoin commented on tests/functional/test_s3.py:
```diff
@@ -11,15 +11,11 @@
 # language governing permissions and limitations under the License.
 from tests import unittest, mock, BaseSessionTest, create_session
-import os
-import nose
```
Yes, its in the commit message.

JordonPhillips commented on botocore/handlers.py:
```diff
@@ -109,6 +110,24 @@ def decode_console_output(parsed, **kwargs):
+    if not input_shape or not hasattr(input_shape, 'members'):
+        return
+    # Might be a mock model from a test that doesn't care about this
+    if not hasattr(input_shape.members.items(), '__iter__'):
```
Why would all tests need that? This should only be running in functional / integration tests aside from the unit tests directly related to it.

JordonPhillips commented on botocore/model.py:
```diff
@@ -408,6 +408,26 @@ def output_shape(self):
 
     @CachedProperty
+    def has_idempotent_members(self):
+        input_shape = self.input_shape
```
It would probably be better to just use `get_idempotent_members` here.

JordonPhillips commented on tests/unit/test_idempotency.py:
```diff
@@ -0,0 +1,32 @@
+            '^[0-9a-f]{8}-([0-9a-f]{4}-){3}[0-9a-f]{12}$',
+            re.I)
+
+    def test_injection(self):
```
You should also add a test for when the param is already provided.

JordonPhillips commented:
I think all that's really missing now is documentation for idempotent parameters.

kyleknap commented on tests/functional/test_ec2.py:
```diff
@@ -0,0 +1,77 @@
+            self.function_name, self.service_response, expected_params)
+
+        with self.stubber:
+            try:
```
I do not think you need this try, except, finally for a couple reasons:

1) The ``self.params_seen`` will always get reset to an empty list because it is defined in ``setUp`` and ``setUp`` gets called for each test.
2) There does not appear to be too much benefit for catching the StubAssertionError and then re-raising it. It seems unneeded.

Let me know what you think.

JordonPhillips replied:
I think this is the last thing that needs to be addressed

kyleknap commented on tests/functional/test_ec2.py:
```diff
@@ -0,0 +1,77 @@
+        # Record all the parameters that get seen
+        self.client.meta.events.register_first(
+            'before-call.*.*',
+            lambda model, params, **kwargs: self.params_seen.extend(
```
I typically try to avoid lambda's as they are harder to debug when things start going wrong. I usually define a function on the TestCase and register the method to the event.

kyleknap replied:
Any update on this?

stealthycoin replied:
Whoops I missed this one...

JordonPhillips commented on tests/functional/test_ec2.py:
```diff
@@ -75,3 +76,10 @@ def test_insert_idempotency_token(self):
                 self.params_seen = []
+
+
+class TestEC2Docs(BaseDocsFunctionalTest):
```
This should be moved to `tests/functional/docs/test_ec2.py`

JordonPhillips commented on botocore/docs/method.py:
```diff
@@ -207,6 +208,19 @@ def document_model_driven_method(section, method_name, operation_model,
+    if 'request-params' in section.available_sections:
+        param_section = section.get_section('request-params')
+        # if operation_model.name == 'PurchaseScheduledInstances':
+        #     pass #import pdb; pdb.set_trace();
```
These comments should be removed

stealthycoin commented:
<img width="782" alt="screen shot 2016-11-29 at 2 45 09 pm" src="https://cloud.githubusercontent.com/assets/1010634/20732252/afd7feac-b642-11e6-8558-badbcd98f85a.png">

kyleknap commented on botocore/docs/method.py:
```diff
@@ -207,6 +208,17 @@ def document_model_driven_method(section, method_name, operation_model,
 
+
+    # Find idempotent params and append documentation
+    if 'request-params' in section.available_sections:
```
Make sure you run flake8. These lines look way too long.

stealthycoin replied:
Yea I had automatic flake8 on all the time and then I turned that off so I forgot to run it manually.

kyleknap commented on botocore/docs/method.py:
```diff
@@ -207,6 +208,17 @@ def document_model_driven_method(section, method_name, operation_model,
+        for param_name in param_section.available_sections:
+            if param_name in context.get('special_shape_types', {}).get('idempotent_members', []):
+                docs = param_section.get_section(param_name).get_section('param-documentation')
+                docs.writeln('This field is autopopulated if not provided.')
```
So based on the example you had. It looks like a block quote, which does not look good. Could remove ``writeln`` and use ``write`` instead to append on the value?

kyleknap commented on botocore/docs/method.py:
```diff
@@ -207,6 +208,17 @@ def document_model_driven_method(section, method_name, operation_model,
+    # Find idempotent params and append documentation
+    if 'request-params' in section.available_sections:
+        param_section = section.get_section('request-params')
+        for param_name in param_section.available_sections:
```
I would prefer we move this in ``params.py`` and ``shape.py``.  It seems weird to be adding parameter specific logic not in one of these files, especially with the use of the context (it seems unneeded right now). You can probably follow the same logic as streaming input shape.

stealthycoin commented:
I changed the wording to match elsewhere in the docs. 
![screen shot 2016-11-30 at 2 37 47 pm](https://cloud.githubusercontent.com/assets/1010634/20774309/a6f8555c-b70a-11e6-8d0e-dbb7607dfd16.png)

Moved everything to `shape.py` and `params.py` I found that the metadata dict is populated with the idempotency property, so it was easier than I thought it would be. Still probably needs to be made more general.

JordonPhillips commented on botocore/docs/shape.py:
```diff
@@ -83,6 +83,17 @@ def traverse_and_document_shape(self, section, shape, history,
+    def _append_special_param_docstring(self, shape, section):
+        special_docs = {
+            'idempotencyToken':
+            'Please note this parameter is automatically populated '
```
This formatting is confusing, and I'm a bit surprised it works. Initially I thought this was a list.

JordonPhillips commented on botocore/docs/shape.py:
```diff
@@ -83,6 +83,17 @@ def traverse_and_document_shape(self, section, shape, history,
+        special_docs = {
+            'idempotencyToken':
+            'Please note this parameter is automatically populated '
+            'if it is not provided. Including this parameter is not required\n'
```
Trailing period is missing

stealthycoin commented:
![screen shot 2016-11-30 at 4 13 58 pm](https://cloud.githubusercontent.com/assets/1010634/20776879/0c788f02-b718-11e6-8b7d-e4789966da20.png)
